src/main/scala/beam/agentsim/infrastructure/HierarchicalParkingManager.scala

```diff
@@ -0,0 +1,290 @@
+package beam.agentsim.infrastructure
+
+import java.time.temporal.ChronoUnit
+import java.time.{ZoneOffset, ZonedDateTime}
+
+import akka.actor.{ActorLogging, ActorRef, Cancellable, Props}
+import akka.event.Logging
+import beam.agentsim.Resource.ReleaseParkingStall
+import beam.agentsim.infrastructure.HierarchicalParkingManager.{ParkingCluster, Worker}
+import beam.agentsim.infrastructure.parking.ParkingZone
+import beam.agentsim.infrastructure.parking.ParkingZoneSearch.ZoneSearchTree
+import beam.agentsim.infrastructure.taz.{TAZ, TAZTreeMap}
+import beam.sim.common.GeoUtils
+import beam.sim.common.GeoUtils.toJtsCoordinate
+import beam.sim.config.BeamConfig
+import com.typesafe.scalalogging.LazyLogging
+import com.vividsolutions.jts.algorithm.ConvexHull
+import com.vividsolutions.jts.geom.prep.{PreparedGeometry, PreparedGeometryFactory}
+import com.vividsolutions.jts.geom.{Coordinate, Envelope, GeometryFactory}
+import de.lmu.ifi.dbs.elki.algorithm.clustering.kmeans.KMeansElkan
+import de.lmu.ifi.dbs.elki.algorithm.clustering.kmeans.initialization.RandomUniformGeneratedInitialMeans
+import de.lmu.ifi.dbs.elki.data.`type`.TypeUtil
+import de.lmu.ifi.dbs.elki.data.{DoubleVector, NumberVector}
+import de.lmu.ifi.dbs.elki.database.StaticArrayDatabase
+import de.lmu.ifi.dbs.elki.database.ids.DBIDIter
+import de.lmu.ifi.dbs.elki.database.relation.Relation
+import de.lmu.ifi.dbs.elki.datasource.ArrayAdapterDatabaseConnection
+import de.lmu.ifi.dbs.elki.distance.distancefunction.minkowski.SquaredEuclideanDistanceFunction
+import de.lmu.ifi.dbs.elki.utilities.random.RandomFactory
+import org.matsim.api.core.v01.{Coord, Id}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.duration._
+import scala.util.Random
+
+/**
+  *
+  * @author Dmitry Openkov
+  */
+class HierarchicalParkingManager(
+  beamConfig: BeamConfig,
+  tazTreeMap: TAZTreeMap,
+  clusters: Vector[ParkingCluster],
+  zones: Array[ParkingZone],
+  searchTree: ZoneSearchTree[TAZ],
+  geo: GeoUtils,
+  rand: Random,
+  boundingBox: Envelope
+) extends beam.utils.CriticalActor
+    with ActorLogging {
+  private var inquiryCounter = 0
+  private var countTimeStart: ZonedDateTime = ZonedDateTime.now(ZoneOffset.UTC)
+  private val tickTask: Cancellable =
+    context.system.scheduler.schedule(2.seconds, 10.seconds, self, "tick")(context.dispatcher)
+
+  private val workers: Vector[Worker] =
+    clusters.zipWithIndex.map {
+      case (cluster, i) =>
+        val emergencyId = Id.create(s"emergency-$i", classOf[TAZ])
+        val actor = context.actorOf(
+          ZonalParkingManager
+            .props(
+              beamConfig,
+              TAZTreeMap.fromSeq(cluster.tazes),
+              zones,
+              searchTree,
+              emergencyId,
+              geo,
+              rand,
```

Comment:
Let's create a local instance of random and pass it to the `ZonalParkingManager` to reduce thread contention:
![image](https://user-images.githubusercontent.com/5107562/80728645-5e4d3d00-8b31-11ea-8cc9-49c7bed1eff7.png)